---
title: Branch length vs HGT number
date: 22 Feb 2018
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: false
    code_folding: hide
    theme: cosmo
---


```{r global_options, include = FALSE}
knitr::opts_chunk$set(	fig.width	= 10, 
						fig.height	= 7, 
						fig.path	= "/Users/aesin/Desktop/Geo_again/HGT_time/Branch_transfer_correlation_220218/Figures/", 
						fig.align	= 'center', 
						dpi			= 300, 
						cache.path	= "/Users/aesin/Desktop/Geo_again/HGT_time/Branch_transfer_correlation_220218/Cache/", 
						warning		= TRUE, 
						message		= TRUE,
						tidy		= TRUE)

```


### Packages & functions
```{r packages, warning = FALSE, message = FALSE}
if (!require("pacman")) install.packages("pacman")
pacman::p_load("data.table", "ape", "phylobase", "phytools", "phangorn", "stringr", "geiger", "RSvgDevice", "reshape2", "ggplot2", "RColorBrewer", "wesanderson", "dplyr")
```

```{r functions, warning = FALSE, message = FALSE}
# Annotate branches function. This is used to label a particular branch with something... #
AnnotateBranch	<- function (edge_entry, index, colName) {
	edge		<- edge_entry[1:2]
	edge_label	<- as.character(round(edge_entry[,colName], digits = 2))
	edgelabels(edge_label, index, adj = c(0.5, -0.25), bg = "white", frame = "none", cex = 1)
}

# / Adjusted plotBranchbyTrait to accept a custom palette / #
plotBranchbyTrait_AE <- function (tree, x, mode = c("edges", "tips", "nodes"), palette = "rainbow", legend = TRUE, xlims = NULL, ...) {
	
	mode <- mode[1]
	if (!inherits(tree, "phylo")) 
		stop("tree should be an object of class \"phylo\".")
	if (mode == "tips") {
		x <- c(x[tree$tip.label], fastAnc(tree, x))
		names(x)[1:length(tree$tip.label)] <- 1:length(tree$tip.label)
		XX <- matrix(x[tree$edge], nrow(tree$edge), 2)
		x <- rowMeans(XX)
	}
	else if (mode == "nodes") {
		XX <- matrix(x[tree$edge], nrow(tree$edge), 2)
		x <- rowMeans(XX)
	}
	if (hasArg(tol)) 
		tol <- list(...)$tol
	else tol <- 1e-06
	if (hasArg(prompt)) 
		prompt <- list(...)$prompt
	else prompt <- FALSE
	if (hasArg(type)) 
		type <- list(...)$type
	else type <- "phylogram"
	if (hasArg(show.tip.label)) 
		show.tip.label <- list(...)$show.tip.label
	else show.tip.label <- TRUE
	if (hasArg(show.node.label)) 
		show.node.label <- list(...)$show.node.label
	else show.node.label <- FALSE
	if (hasArg(edge.width)) 
		edge.width <- list(...)$edge.width
	else edge.width <- 4
	if (hasArg(edge.lty)) 
		edge.lty <- list(...)$edge.lty
	else edge.lty <- 1
	if (hasArg(font)) 
		font <- list(...)$font
	else font <- 3
	if (hasArg(cex)) 
		cex <- list(...)$cex
	else cex <- par("cex")
	if (hasArg(adj)) 
		adj <- list(...)$adj
	else adj <- NULL
	if (hasArg(srt)) 
		srt <- list(...)$srt
	else srt <- 0
	if (hasArg(no.margin)) 
		no.margin <- list(...)$no.margin
	else no.margin <- TRUE
	if (hasArg(root.edge)) 
		root.edge <- list(...)$root.edge
	else root.edge <- FALSE
	if (hasArg(label.offset)) 
		label.offset <- list(...)$label.offset
	else label.offset <- 0.01 * max(nodeHeights(tree))
	if (hasArg(underscore)) 
		underscore <- list(...)$underscore
	else underscore <- FALSE
	if (hasArg(x.lim)) 
		x.lim <- list(...)$x.lim
	else x.lim <- NULL
	if (hasArg(y.lim)) 
		y.lim <- list(...)$y.lim
	else y.lim <- if (legend && !prompt && type %in% c("phylogram", 
		"cladogram")) 
		c(1 - 0.06 * length(tree$tip.label), length(tree$tip.label))
	else NULL
	if (hasArg(direction)) 
		direction <- list(...)$direction
	else direction <- "rightwards"
	if (hasArg(lab4ut)) 
		lab4ut <- list(...)$lab4ut
	else lab4ut <- NULL
	if (hasArg(tip.color)) 
		tip.color <- list(...)$tip.color
	else tip.color <- "black"
	if (hasArg(plot)) 
		plot <- list(...)$plot
	else plot <- TRUE
	if (hasArg(rotate.tree)) 
		rotate.tree <- list(...)$rotate.tree
	else rotate.tree <- 0
	if (hasArg(open.angle)) 
		open.angle <- list(...)$open.angle
	else open.angle <- 0
	if (is.function(palette)) 
		cols <- palette(n = 1000)
	else {
		if (palette == "heat.colors") 
			cols <- heat.colors(n = 1000)
		if (palette == "gray") 
			cols <- gray(1000:1/1000)
		if (palette == "rainbow") 
			cols <- rainbow(1000, start = 0.7, end = 0)
		else
			cols <- palette
	}
	if (is.null(xlims)) 
		xlims <- range(x) + c(-tol, tol)
	breaks <- 0:1000/1000 * (xlims[2] - xlims[1]) + xlims[1]
	whichColor <- function(p, cols, breaks) {
		i <- 1
		while (p >= breaks[i] && p > breaks[i + 1]) i <- i + 
			1
		cols[i]
	}
	colors <- sapply(x, whichColor, cols = cols, breaks = breaks)
	par(lend = 2)
	xx <- plot.phylo(tree, type = type, show.tip.label = show.tip.label, 
		show.node.label = show.node.label, edge.color = colors, 
		edge.width = edge.width, edge.lty = edge.lty, font = font, 
		cex = cex, adj = adj, srt = srt, no.margin = no.margin, 
		root.edge = root.edge, label.offset = label.offset, underscore = underscore, 
		x.lim = x.lim, y.lim = y.lim, direction = direction, 
		lab4ut = lab4ut, tip.color = tip.color, plot = plot, 
		rotate.tree = rotate.tree, open.angle = open.angle, lend = 2, 
		new = FALSE)
	if (legend == TRUE && is.logical(legend)) 
		legend <- round(0.3 * max(nodeHeights(tree)), 2)
	if (legend) {
		if (hasArg(title)) 
			title <- list(...)$title
		else title <- "trait value"
		if (hasArg(digits)) 
			digits <- list(...)$digits
		else digits <- 1
		if (prompt) 
			add.color.bar(legend, cols, title, xlims, digits, 
				prompt = TRUE)
		else add.color.bar(legend, cols, title, xlims, digits, 
			prompt = FALSE, x = par()$usr[1] + 0.05 * (par()$usr[2] - 
				par()$usr[1]), y = par()$usr[3] + 0.05 * (par()$usr[4] - 
				par()$usr[3]))
	}
	invisible(xx)
}
```


# Input data {.tabset .tabset-fade}
## Read in tables
```{r readInLists, warning = FALSE, message = FALSE, cache = TRUE}

penalty_list	<- list(4, 5, 6)

# Define the working directories
main_dir		<- "/Users/aesin/Desktop/Geo_again"
genomes_dir		<- file.path(main_dir, "Genomes")
mowgli_dir		<- file.path(main_dir, "Mowgli")
output_dir		<- file.path(main_dir, "HGT_time", "Data")

# Create output directory
dir.create(output_dir, showWarnings = FALSE)

# This directory contains the consistent HGT results
# AE: changed Refined_events to Clean_events to remove the plasmid-contaminated events
refinedHGTs_dir	<- file.path(mowgli_dir, "Mowgli_output", "Cleaned_events", "HGT_events")

# Define and read in the list of all AG tips as they appear in the Mowgli species tree
AG_mowCladoTips_file	<- file.path(mowgli_dir, "Inside_group", "AnoxyGeo_mowTips.txt")
AG_mowCladoTips_char	<- read.table(file = AG_mowCladoTips_file, sep = "\n", stringsAsFactors = FALSE)$V1

# Define and read in the taxid <-> binomial translation table
taxidBinom_trans_file	<- file.path(genomes_dir, "Genome_lists", "Taxid_refinedBinomial_table.tsv")
taxidBinom_trans_df		<- read.table(file = taxidBinom_trans_file, sep = "\t", header = TRUE,stringsAsFactors = FALSE)

```

## Read in trees
```{r readInTrees, warning = FALSE, message = FALSE, cache = TRUE}
# Define and read in the mowgli species tree
mowSpeciesTree_file	<- file.path(mowgli_dir, "Mowgli_output", "Output_3", "1", "outputSpeciesTree.mpr")
if (!file.exists(mowSpeciesTree_file)) stop("Cannot find Mowgli species tree file, try a different file")
mowSpecies_tree		<- read.tree(mowSpeciesTree_file)

# Define and read in the consensus time tree (same topology as mowgli tree)
consensusTree_file	<- file.path(main_dir, "Consensus_groups", "Bacillaceae", "Final_trees", "Taxid_labelled", "RAxML_bipartitions.super_tree-50.txt")
consensusTree_tree	<- read.tree(file = consensusTree_file)
```



# Clado and Time trees {.tabset .tabset-fade}
## Clado-Mowgli tree
```{r processCladoTree, warning = FALSE, message = FALSE, cache = TRUE}
# Using the AG mowgli tips labels extract the AG subtree
AG_mowClado_tree		<- drop.tip(mowSpecies_tree, setdiff(mowSpecies_tree$tip.label, AG_mowCladoTips_char))

# Produce a df with corresponding taxid - mowgli tip entries
taxidMowExtend_df			<- data.frame(Full = AG_mowClado_tree$tip.label, str_split(AG_mowClado_tree$tip.label, "_", simplify = TRUE), stringsAsFactors = FALSE)
# Final column is the binomials
taxidMowExtend_df[,4]		<- apply(taxidMowExtend_df, 1, function(row) {
	taxid	<- row[2]
	binom	<- taxidBinom_trans_df$Binomial[which(taxidBinom_trans_df$Taxid == taxid)]
})
# Rename columns 
names(taxidMowExtend_df)	<- c("Full", "Taxid", "Extension", "Binomial")

## Write out the table for future use
write.table(taxidMowExtend_df, file = file.path(output_dir, "Taxid2MowTip_table.tsv"), sep = "\t", quote = FALSE, row.names = FALSE)

# Translate the mowgli clado tree tips to binomials
AG_mowCladoBin_tree		<- AG_mowClado_tree
AG_mowCladoBin_tree$tip.label	<- unlist(lapply(AG_mowCladoBin_tree$tip.label, function(tip) {
	newTipLabel_char	<- taxidMowExtend_df$Binomial[which(taxidMowExtend_df$Full == tip)]
}))

# Translate the mowgli clado tree tips to the mowgli node IDs
AG_mowClado_tree$tip.label	<- unlist(lapply(AG_mowClado_tree$tip.label, function(tip) {
	newTipLabel_char	<- taxidMowExtend_df$Extension[which(taxidMowExtend_df$Full == tip)]
}))

# Transform to phylo4 object
AG_mowClado_as4	<- phylo4(AG_mowClado_tree)
```


## Time-resolved tree
```{r processTimeTree, message = FALSE, warning = FALSE, cache = TRUE}
# Extract the AG subtree
AG_conTime_tree		<- drop.tip(consensusTree_tree, setdiff(consensusTree_tree$tip.label, taxidMowExtend_df$Taxid))

# Relabel the time tree to binomials
AG_conTimeBin_tree	<- AG_conTime_tree
AG_conTimeBin_tree$tip.label	<- unlist(lapply(AG_conTimeBin_tree$tip.label, function(tip) {
	newTipLabel_char		<- taxidMowExtend_df$Binomial[which(taxidMowExtend_df$Taxid == tip)]
}))

# Relabel the time tree to the Mowgli extension
AG_conTime_tree$tip.label	<- unlist(lapply(AG_conTime_tree$tip.label, function(tip) {
	newTipLabel_char		<- taxidMowExtend_df$Extension[which(taxidMowExtend_df$Taxid == tip)]
}))

# Remove the BS values and conver the phylo4
AG_conTime_tree$node.label	<- NULL
AG_conTime_as4		<- phylo4(AG_conTime_tree)


# Double check the clado (Astral) and time (consensus) topologies are the same
cladoVsTime_rfDist	<- RF.dist(AG_mowClado_tree, AG_conTime_tree)
```

The Robinson-Foulds distance between the Mowgli (Clado) and Consensus (Time) trees is: `r cladoVsTime_rfDist`


## Confirm node identity
```{r matchCladoTimeNodes, message = FALSE, warning = FALSE, cache = TRUE}
# Relabel internal timeTree nodes to match the Mowgli labels in the clado tree
AG_conTime_intNodes		<- nodeId(AG_conTime_as4, "internal")

# Get the node labels as a data frame
AG_mowClado_intNodes	<- nodeId(AG_mowClado_as4, "internal")

# For each internal node in the time tree find a corresponding internal clado
# node with the same underlying tips, then relabel that time tree node with the
# corresponding clado (mowgli) label
for (timeIntNode in AG_conTime_intNodes) {
	matched	<- FALSE
	for (cladoIntNode in AG_mowClado_intNodes) {
		if (setequal(tips(AG_conTime_tree, timeIntNode), tips(AG_mowClado_tree, cladoIntNode)) == TRUE) {

			## Get the label of the node from the species-cladogram tree ##
			node_label	<- AG_mowClado_as4@label[cladoIntNode]

			## Index of the internal node label in the time-resolved tree ##
			node_index	<- which(names(nodeLabels(AG_conTime_as4)) == timeIntNode)
			
			## Assign the name of the node to the time-resolved tree ##
			nodeLabels(AG_conTime_as4)[node_index] <- node_label

			cat(paste0("Time-resolved tree node: ", timeIntNode, " maps to species tree node: ", cladoIntNode, "\n"))
			matched	<- TRUE
			break
		}
	}
	if (matched == FALSE) {
		stop(paste0("Time tree node: ", timeIntNode, " did not map to any species tree node\n"))
	}
}

## Write out the time tree (for use in other scripts - e.g. HGT_position)
write.tree(as(AG_conTime_as4, "phylo"), file = file.path(output_dir, "AG_conTimeTree.tree"))
```

# Read in HGT {.tabset .tabset-fade}
## Prepare output tables
```{r prepNodeEdgeDfs, message = FALSE, warning = FALSE, cache = TRUE}
# A list of two dfs - the lists of all nodes found in the Clado and Time trees
# These will be matched to the receptor nodes of the HGTs below, and is why we 
# needed to relabel the tips to the Mowgli node IDs - to match transfer nodes.
allNodeLabels_dfs	<- list(
	clado = data.frame(NodeLabels = labels(AG_mowClado_as4, "all"), stringsAsFactors = FALSE),
	time = data.frame(NodeLabels = labels(AG_conTime_as4, "all"), stringsAsFactors = FALSE)
)

# Prepare tree-edge based dfs. Each edge, flanked by two nodes (E1 & E2) will
# have transfers assigned. In the HGT prediction the "receptor node" corresponds
# to the bottom of the edge (E2). The list of node (above) and edges appear in
# an identical order within each tree, so we can assign the transfers to the edges
# using the mowgli node IDs.
allEdgeFinal_dfs	<- list(clado = data.frame(AG_mowClado_tree$edge, Subgroup = NA), time = data.frame(AG_conTime_tree$edge, Subgroup = NA))
allEdgeFinal_dfs	<- lapply(allEdgeFinal_dfs, setnames, c("E1", "E2", "Subgroup"))

# Create columns for each penalty tested (minimal penalty for consistent HGT)
# is T4 (i.e. consistent over penalties 3 and 4)
allPenNames_char	<- lapply(penalty_list, function(penalty) paste0("T", penalty))
allEdgeFinal_dfs	<- lapply(allEdgeFinal_dfs, function(df) {
	for (penName in allPenNames_char) {0 -> df[eval(penName)]}
	return(df)
})
```

## Process lHGT data
```{r countHGTbyEdge, message = FALSE, warning = FALSE, cache = TRUE}
# For each penalty, process the lHGTs into each node, and assign them to an edge
# in the tree-edge based dfs
for (penalty in penalty_list) {

	# Define and read in the consistent lHGT file
	lHGT_name_file	<- paste0("T", penalty, "_full_lHGT_events.tsv")
	lHGT_data_df	<- read.table(file = file.path(refinedHGTs_dir, lHGT_name_file), header = TRUE, sep = "\t")

	# Define column to which HGTs are assigned
	colName			<- paste0("T", penalty)
	# Counter for transfers at the AG root (not plotted on tree)
	at_root			<- 0

	# Process the lHGT results line-by-line
	for (rowIndex in 1:nrow(lHGT_data_df)) {

		# Define line and extract the "receptor node"
		row 		<- lHGT_data_df[rowIndex,]
		recNodes	<- str_split(row$Receptor_nodes, pattern = " ", simplify = T)

		# Check whether transfer is into the root of AG
		if (recNodes[2] == 1375) {
			at_root	<- at_root + 1
		# This would be a transfer above the root of AG and shouldn't happen
		} else if (recNodes[2] == 1531) {
			message("Should not get this")
		} else {
			# Pull out row number (indication of phylogeny edge) corresponding to the Mowgli tree label.
			edgeForClado	<- grep(recNodes[2], allNodeLabels_dfs$clado$NodeLabels)
			edgeForTime		<- grep(recNodes[2], allNodeLabels_dfs$time$NodeLabels)

			# Shortcut names for the corresponding edge + penalty values (ongoing count)
			cladoEntry		<- allEdgeFinal_dfs$clado[[colName]][which(allEdgeFinal_dfs$clado$E2 == edgeForClado)]
			timeEntry		<- allEdgeFinal_dfs$time[[colName]][which(allEdgeFinal_dfs$time$E2 == edgeForTime)]

			# Update both Time and Clado trees with an extra transfer at the edge + penalty
			allEdgeFinal_dfs$clado[[colName]][which(allEdgeFinal_dfs$clado$E2 == edgeForClado)]	<-  cladoEntry + 1      	
			allEdgeFinal_dfs$time[[colName]][which(allEdgeFinal_dfs$time$E2 == edgeForTime)]	<-  timeEntry + 1
		}
	}
	message(paste0("Penalty: " , penalty, " - at root transfers: ", at_root))
}
```

# Subdividing the _Anoxy / Geobacillus_ clade {.tabset .tabset-fade}
_Aliyu et al 2016_ divides the _Geobacillus_ phylogeny into a number of species subgroups (fig 3). I applied that classification to the genomes used in this analysis. Branches separating species/tips belonging to the same subgroup are considered to not have undergone substantial differentiation and purifying selection.

## Read in subdivisions
```{r prepSubgroupData, message = FALSE, warning = FALSE, cache = TRUE}

# Read in the table constructed based on the publication #
subspeciesGroup_file	<- file.path(genomes_dir, "Genome_lists", "AG_subspeciesGroups.txt")
subspeciesGroup_data	<- read.table(file = subspeciesGroup_file, header = TRUE, sep = "\t", stringsAsFactors = FALSE)

# Get those tips that have a fellow tip in the same subgroup. Isolate all the subgroups that have more than one member #
subspecOnly_data		<- subset(subspeciesGroup_data, duplicated(subspeciesGroup_data$Group) | duplicated(subspeciesGroup_data$Group, fromLast = TRUE))
uniqueSubGroup_char		<- unique(subspecOnly_data$Group)
```

## Assign subgroup identity
```{r assignSubgroup, message = FALSE, warning = FALSE, cache = TRUE}
# Label branches that connect tips belonging to the same subgroup:
# For each group with more than one member, identify the tips. Find all the
# descendant nodes from the common ancestor of the subgroup (they are
# monophyletic). Using those nodes, identify all the edges (branches) that we
# will classify as being a subgroup branch (1). All other branches are separate
# groups (0)
for (subGroup in uniqueSubGroup_char) {

	taxidsInSubGroup	<- as.vector(subspecOnly_data$Taxid[which(subspecOnly_data$Group == subGroup)])
	taxidsToMowTips		<- taxidMowExtend_df$Extension[which(taxidMowExtend_df$Taxid %in% taxidsInSubGroup)]
	allSubGroupNodes	<- as.vector(descendants(AG_conTime_as4, MRCA(AG_conTime_as4, taxidsToMowTips), "all"))

	for (node in allSubGroupNodes) {
		allEdgeFinal_dfs$time$Subgroup[which(allEdgeFinal_dfs$time$E2 == node)]		<- 1
		allEdgeFinal_dfs$clado$Subgroup[which(allEdgeFinal_dfs$clado$E2 == node)]	<- 1
	}
}

# All other edges labelled as 0 (not subgroup)
allEdgeFinal_dfs	<- lapply(allEdgeFinal_dfs, function(df) {
	df$Subgroup[is.na(df$Subgroup)]	<- 0
	return(df)
})
```

```{r bySubgroupTimeTree_plot, message = FALSE, warning = FALSE, cache = TRUE, fig.width = 10, fig.height = 6}
Royal1Pal	<- colorRampPalette(rev(wes_palette("Royal1")[1:2]))
plotBranchbyTrait_AE(AG_conTimeBin_tree, allEdgeFinal_dfs$time$Subgroup, method = "edges", palette = Royal1Pal, title = "Subgroup branches")
```


# HGT per branch {.tabset .tabset-fade}
## Calculate fraction of all HGT
```{r averaging_transfers, message = FALSE, warning = FALSE, cache = TRUE}
allEdgeFinal_fract_dfs	<- lapply(allEdgeFinal_dfs, function(df) cbind(df[,1:3], sweep(df[,-1:-3] , 2, colSums(df[,-1:-3]), "/") * 100))
allEdgeFinal_fract_dfs	<- lapply(allEdgeFinal_fract_dfs, function(df) cbind(df, Mean = rowMeans(df[,-1:-3])))
```

## HGT-per-branch - Clado
```{r cladoTreeByHGT_plot, message = FALSE, warning = FALSE, cache = TRUE, fig.width = 12, fig.height = 7}
# Color palette for HGT / branch 
conPal <- colorRampPalette(wes_palette("Zissou1"))

# Plot mean fraction of transfer per branch on the Clado tree
plotBranchbyTrait_AE(AG_mowCladoBin_tree, allEdgeFinal_fract_dfs$clado$Mean, method = "edges", legend = 5, palette = conPal, title = "Fraction of transfers")

# Anotate the branches with the numeric values
for (index in 1:nrow(allEdgeFinal_fract_dfs$clado)) {
	entry <- allEdgeFinal_fract_dfs$clado[index,]
	AnnotateBranch(entry, index, "Mean")
}
```

## HGT-per-branch - Time
```{r timeTreeByHGT_plot, message = FALSE, warning = FALSE, cache = TRUE, fig.width = 12, fig.height = 7}
# Plot mean fraction of transfer per branch on the Time tree
plotBranchbyTrait_AE(AG_conTimeBin_tree, allEdgeFinal_fract_dfs$time$Mean, method = "edges", legend = .1, palette = conPal, title = "Fraction of transfers")

# Anotate the branches with the numeric values
for (index in 1:nrow(allEdgeFinal_fract_dfs$time)) {
	entry <- allEdgeFinal_fract_dfs$time[index,]
	AnnotateBranch(entry, index, "Mean")
}
```


# Continuous HGT

## HGT vs Branch Length
1. Add the branch lengths from the time-resolved tree to the time-resolved dataframe. We take these directly from the Time-resolved tree: [2. Time-resolved tree](#sect1). At same time, add a point index, a mapping for each branch so any point in the correlation point can be easily associated with a branch on the tree.
``` {r BranchLengthAdd, message = FALSE, warning = FALSE, cache = TRUE}
allEdgeFinal_dfs$time	<- data.frame(
	allEdgeFinal_dfs$time,
	Mean = allEdgeFinal_fract_dfs$time$Mean,
	BranchLen = AG_conTime_tree$edge.length,
	Index = rownames(allEdgeFinal_dfs$time),
	SubGroupLogic = as.logical(allEdgeFinal_dfs$time$Subgroup),
	stringsAsFactors = FALSE
)
```
```{r branchLengthVsHGT_dotPlot, message = FALSE, warning = FALSE, cache = TRUE, fig.width = 8, fig.height = 8}
branchVsHGT_dotplot	<- ggplot(data = allEdgeFinal_dfs$time, mapping = aes(x = BranchLen, y = Mean, label = Index, color = SubGroupLogic)) +
	geom_point(size = 3) +
	geom_text(aes(label = Index), hjust = 0, vjust = -1) +
	xlab("Branch length") + ylab("Mean fraction of HGTs across penalties") +
	scale_color_manual(values = rev(wes_palette("Royal1")[1:2])) +
	theme(panel.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.line = element_line(colour = "black"), panel.grid.major = element_line(colour = "grey80"))
plot(branchVsHGT_dotplot)
```

```{r timeTreeByGroupIndex_plot, message = FALSE, warning = FALSE, cache = TRUE, fig.width = 10, fig.height = 6}
Royal1Pal	<- colorRampPalette(rev(wes_palette("Royal1")[1:2]))
plotBranchbyTrait_AE(AG_mowCladoBin_tree, allEdgeFinal_dfs$time$Subgroup, method = "edges", palette = Royal1Pal, title = "Subgroup branches")
for (i in as.vector(allEdgeFinal_dfs$time$Index)) {
	edgelabels(i, as.numeric(i), adj = c(0.5, -0.25), bg = "white", frame = "none", cex = 0.8)
}
```


## HGT vs Branch Length: Correlation
```{r crossPenaltyCorCalc, message = FALSE, warning = FALSE, cache = TRUE}

perPenaltyCor_list	<- lapply(c(penalty_list, "Mean"), function(penalty) {
	
	if (is.numeric(penalty)) {
		penColName	<- paste0("T", penalty)
	} else {
		penColName	<- penalty
	}
	time_df			<- allEdgeFinal_dfs$time

	# Overall pearson + rsquared
	all_Pcor		<- cor(time_df$BranchLen, time_df[[penColName]])
	allRsqrd		<- summary(lm(time_df$BranchLen ~ time_df[[penColName]]))$r.squared

	# Pearson + rsquared (without two obvious outliers)
	outlierIndex	<- as.character(c(1, 8))
	timeNoOut_df	<- time_df[-which(time_df$Index %in% outlierIndex),]

	noOut_Pcor		<- cor(timeNoOut_df$BranchLen, timeNoOut_df[[penColName]])
	noOutRsqrd		<- summary(lm(timeNoOut_df$BranchLen ~ timeNoOut_df[[penColName]]))$r.squared

	## Correlation when subgroup removed
	grpTime_df		<- time_df[which(time_df$SubGroupLogic == FALSE),]
	# Pearson + rsquared
	grp_Pcor		<- cor(grpTime_df$BranchLen, grpTime_df[[penColName]])
	grpRsqrd		<- summary(lm(grpTime_df$BranchLen ~ grpTime_df[[penColName]]))$r.squared

	# Pearson + rsquared (without two obvious outliers)
	grpTimeNoOut_df	<- grpTime_df[-which(grpTime_df$Index %in% outlierIndex),]
	grpNoOut_Pcor	<- cor(grpTimeNoOut_df$BranchLen, grpTimeNoOut_df[[penColName]])
	grpNoOutRsqrd	<- summary(lm(grpTimeNoOut_df$BranchLen ~ grpTimeNoOut_df[[penColName]]))$r.squared

	outputPearson	<- data.frame(Penalty = as.character(penalty), Test = "Pearson", All = all_Pcor, NoOutlier = noOut_Pcor, Group = grp_Pcor, GroupNoOutlier = grpNoOut_Pcor, stringsAsFactors = FALSE)
	outputRsqrd		<- data.frame(Penalty = as.character(penalty), Test = "Rsquared", All = allRsqrd, NoOutlier = noOutRsqrd, Group = grpRsqrd, GroupNoOutlier = grpNoOutRsqrd, stringsAsFactors = FALSE)

	return(rbind(outputPearson, outputRsqrd))
})

perPenaltyCor_df	<- bind_rows(perPenaltyCor_list)
perPenaltyCor_melt	<- melt(perPenaltyCor_df, id.vars = c("Penalty", "Test"), variable.name = "Subset", value.name = "StatisticValue")
perPenaltyCor_melt$Penalty	<- factor(as.character(perPenaltyCor_melt$Penalty), levels = unique(as.character(perPenaltyCor_melt$Penalty)))
```

```{r crossPenaltyCor_plot, message = FALSE, warning = FALSE, cache = TRUE, fig.width = 8, fig.height = 8}
perPenaltyCor_plot	<- ggplot(data = perPenaltyCor_melt, mapping = aes(x = Penalty, y = StatisticValue, color = Subset, group = Subset)) +
	geom_point(size = 3) +
	geom_line(size = 2) +
	scale_y_continuous(limits = c(0, 1)) +
	facet_wrap(~Test) +
	scale_color_manual(values = rev(wes_palette("BottleRocket2")[1:4])) +
	theme(panel.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.line = element_line(colour = "black"), panel.grid.major = element_line(colour = "grey80"))
plot(perPenaltyCor_plot)
```
















